### PR TITLE
Fix #4338: Less mutability in CoreJSLib

### LINKED ISSUE #4338: Reconsider mutability in CoreJSLib
See https://github.com/scala-js/scala-js/pull/4321#r545197842

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

**Comment by gzm0** [comment]
@sjrd this is somewhat what I had in mind. An issue that crops up here (not the only place in our codebase IMHO), is that we do not have a way to write lists without large (4 spcs) indentation. So in this PR, some of the indentation is wrong :-/ Of course I can just indent it more, but IMO that hurts readability at some point.

**Comment by sjrd** [comment]
Yes, that is a problem. In a sense I would appeal to the "DSL override" rule: when we're writing JS ASTs by hand, we're essentially using a DSL, and hence all coding style rules can be adapted to match that DSL. With that in mind, we could declare that, within such code blocks, indentation for `Block()`s (and others) that use a new line for each argument should use 2 spaces only:
```scala
Block(
  foo,
  bar
)
```
The problem is that we can never teach that to scalafmt, if we ever manage to apply it to this codebase.

**Comment by gzm0** [comment]
The only alternatives to simply allowing it I see are:

- Add braces (this pattern seems to already be sometimes in use).
    ```
    Block({
     foo
    }, {
     bar
    })
    ```

    But that's really only viable if the bodies are longer, otherwise it's a lot of clutter (the again, with shorter bodies, 4 spaces of indent is not a problem).

- Start distinguishing between binpacked parenthesized lists and per-line (gofmt uses this IMO very successfully). So the following styles would be allowed:
    ```
    Block(foo, bar,
        baz)

    Block(
      foo,
      bar, // <-- trailing comma is how gofmt distinguishes the two, we probably can't.
    )
    ```
   
   This would be a change in our coding style of course.

**Comment by sjrd** [comment]
> Start distinguishing between binpacked parenthesized lists and per-line (gofmt uses this IMO very successfully). So the following styles would be allowed:

That actually looks like a very reasonable coding style. We could do that.

We can't use the trailing comma as indicator (yet) but we can use the fact that the initial opening `(` is directly followed by a newline. That's good as well.

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala [hunk 1]
```diff
@@ -731,278 +694,244 @@ private[emitter] object CoreJSLib {
...
+                }, {
+                  Apply(BigIntRef,
+                      Apply(genIdentBracketSelect(MathRef, "trunc"), x :: Nil) :: Nil)
+                })
               })
             })
-          })
-        })
-      }
+          }}
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:798]: Hum, there's no precedent for for having a double `}}`, ever. I'm not even sure we can have a closing brace anywhere but at the start of a line (except in `import`s). I see that this comes from
```scala
defineFunction1("doubleToLong") { x => Return {
```
but I don't think we should allow this. A lambda formatted with `{ params =>` should always have its body start on the next line.

IMO there are two solutions here. Either put `Return {` on the next line, and indent everything one level more. Or format it as
```scala
defineFunction1("doubleToLong")(x => Return {
  ...
})
```

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala [hunk 2]
```diff
@@ -731,278 +694,244 @@ private[emitter] object CoreJSLib {
...
+
+          // This basically emulates the ECMAScript specification for 'new'.
+          Block(
+              const(instance, Apply(genIdentBracketSelect(ObjectRef, "create"), ctor.prototype :: Nil)),
+              const(result, Apply(genIdentBracketSelect(ctor, "apply"), instance :: args :: Nil)),
+              Switch(typeof(result),
+                  List("string", "number", "boolean", "undefined").map(str(_) -> Skip()) :+
+                  str("symbol") -> Return(instance),
+                  Return(If(result === Null(), instance, result)))
+          )
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:817]:
```suggestion
          Block(
            const(instance, Apply(genIdentBracketSelect(ObjectRef, "create"), ctor.prototype :: Nil)),
            const(result, Apply(genIdentBracketSelect(ctor, "apply"), instance :: args :: Nil)),
            Switch(typeof(result),
                List("string", "number", "boolean", "undefined").map(str(_) -> Skip()) :+
                str("symbol") -> Return(instance),
                Return(If(result === Null(), instance, result)))
          )
```

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala [hunk 3]
```diff
@@ -731,278 +694,244 @@ private[emitter] object CoreJSLib {
...
+            const(getOwnPropertyDescriptor, genIdentBracketSelect(ObjectRef, "getOwnPropertyDescriptor")),
+            let(superProto, superClass.prototype),
+            While(superProto !== Null(), {
+              Block(
+                  const(desc, Apply(getOwnPropertyDescriptor, superProto :: propName :: Nil)),
+                  If(desc !== Undefined(), Return(desc), Skip()),
+                  superProto := Apply(getPrototypeOf, superProto :: Nil)
+              )
+            })
+        )
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:838]:
```suggestion
        Block(
          const(getPrototypeOf, genIdentBracketSelect(ObjectRef, "getPrototyeOf")),
          const(getOwnPropertyDescriptor, genIdentBracketSelect(ObjectRef, "getOwnPropertyDescriptor")),
          let(superProto, superClass.prototype),
          While(superProto !== Null(), Block(
            const(desc, Apply(getOwnPropertyDescriptor, superProto :: propName :: Nil)),
            If(desc !== Undefined(), Return(desc), Skip()),
            superProto := Apply(getPrototypeOf, superProto :: Nil)
          ))
        )
```

Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:838]: Several other functions below can be reformatted as well using the same scheme. They've all got their indentation affected anyway, so we might as well use the opportunity to benefit from the new style.

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala [hunk 4]
```diff
@@ -731,278 +694,244 @@ private[emitter] object CoreJSLib {
...
-              If((srcArray !== destArray) || (destPos < srcPos) || (((srcPos + length) | 0) < destPos), {
-                For(let(i, 0), i < length, i := ((i + 1) | 0), {
-                  BracketSelect(destArray, (destPos + i) | 0) := BracketSelect(srcArray, (srcPos + i) | 0)
-                })
-              }, {
-                For(let(i, (length - 1) | 0), i >= 0, i := ((i - 1) | 0), {
-                  BracketSelect(destArray, (destPos + i) | 0) := BracketSelect(srcArray, (srcPos + i) | 0)
-                })
+        Block(
+            if (arrayIndexOutOfBounds != CheckedBehavior.Unchecked) {
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:912]: Here we can decrease the indentation as well.

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala [hunk 5]
```diff
@@ -1076,60 +1000,68 @@ private[emitter] object CoreJSLib {
...
 
-        if (useECMAScript2015) {
-          val f = weakMapBasedFunction
-          defineFunction("systemIdentityHashCode", f.args, f.body)
-        } else {
-          buf += extractWithGlobals(globalVarDef("systemIdentityHashCode", CoreVar,
-              If(idHashCodeMap !== Null(), weakMapBasedFunction, fieldBasedFunction)))
-        }
+        Block(
+            let(lastIDHash, 0),
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:1004]:
```suggestion
        Block(
          let(lastIDHash, 0),
```

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala [hunk 6]
```diff
@@ -1076,60 +1000,68 @@ private[emitter] object CoreJSLib {
...
+        val v = varRef("v")
 
-      def defineIsIntLike(name: String, specificTest: Tree): Unit = {
-        defineFunction(name, paramList(v), {
-          Return((typeof(v) === str("number")) && specificTest &&
-              ((int(1) / v) !== (int(1) / double(-0.0))))
-        })
-      }
+        def defineIsIntLike(name: String, specificTest: Tree): Tree = {
+          defineFunction(name, paramList(v), {
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:1024]:
```suggestion
          defineFunction1(name) { v =>
```
and therefore remove the `val v`?

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala [hunk 7]
```diff
@@ -1076,60 +1000,68 @@ private[emitter] object CoreJSLib {
...
 
       if (asInstanceOfs != CheckedBehavior.Unchecked) {
         // Unboxes for everything
-        def defineUnbox(name: String, boxedClassName: ClassName, resultExpr: Tree): Unit = {
+
+        val v = varRef("v")
+
+        def defineUnbox(name: String, boxedClassName: ClassName, resultExpr: Tree): Tree = {
           val fullName = boxedClassName.nameString
           defineFunction(name, paramList(v), Return {
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:1066]: Same here:
```suggestion
          defineFunction1(name)(v => Return {
```

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala [hunk 8]
```diff
@@ -1249,22 +1185,27 @@ private[emitter] object CoreJSLib {
...
+          extractWithGlobals(globalClassDef("ac", componentTypeRef,
               Some(globalVar("c", ObjectClass)), ctor :: members))
         } else {
-          buf += extractWithGlobals(globalFunctionDef("ac", componentTypeRef,
-              ctor.args, ctor.body))
-          buf += (ArrayClass.prototype := New(globalVar("h", ObjectClass), Nil))
-          buf += (ArrayClass.prototype DOT "constructor" := ArrayClass)
-          buf += assignES5ClassMembers(ArrayClass, members)
+          val clsDef = Block(
+              extractWithGlobals(globalFunctionDef("ac", componentTypeRef,
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:1193]:
```suggestion
          val clsDef = Block(
            extractWithGlobals(globalFunctionDef("ac", componentTypeRef,
```

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala [hunk 9]
```diff
@@ -1249,22 +1185,27 @@ private[emitter] object CoreJSLib {
...
+              (ArrayClass.prototype DOT "constructor" := ArrayClass),
+              assignES5ClassMembers(ArrayClass, members)
+          )
 
           componentTypeRef match {
             case _: ClassRef =>
-              buf += extractWithGlobals(globalFunctionDef("ah", ObjectClass, Nil, Skip()))
-              buf += (globalVar("ah", ObjectClass).prototype := ArrayClass.prototype)
+              Block(
+                  clsDef,
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:1203]:
```suggestion
              Block(
                clsDef,
```

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala [hunk 10]
```diff
@@ -1764,43 +1709,42 @@ private[emitter] object CoreJSLib {
...
-        buf += publicFieldSet("name", str(fullName))
-        buf += publicFieldSet("isInstance",
-            genArrowFunction(paramList(obj), Return(obj !== Null())))
-        buf += privateFieldSet("_arrayOf", {
-          Apply(New(globalVar("TypeData", CoreVar), Nil) DOT "initSpecializedArray", List(
-              typeDataVar,
-              globalVar("ac", ObjectClass),
-              Undefined(), // typedArray
+        Block(
+            extractWithGlobals(
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:1713]:
```suggestion
        Block(
          extractWithGlobals(
```

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala [hunk 11]
```diff
@@ -1830,10 +1774,39 @@ private[emitter] object CoreJSLib {
...
+
+      Block(obj :: prims)
     }
 
-    private def defineFunction(name: String, args: List[ParamDef], body: Tree): Unit =
-      buf += extractWithGlobals(globalFunctionDef(name, CoreVar, args, body))
+    private def defineFunction(name: String, args: List[ParamDef], body: Tree): Tree =
+      extractWithGlobals(globalFunctionDef(name, CoreVar, args, body))
+
+    private val argRefs = Stream.from(1).map(i => varRef("arg" + i))
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:1784]: `Stream` is deprecated in 2.13, so we can't use it. Anyway, it seems to me that the laziness is not really necessary, and that using a List with a fixed size would be more straightforward and efficient:
```suggestion
    private val argRefs = List.tabulate(5)(i => varRef("arg" + i))
```

**Review by sjrd — CHANGES_REQUESTED** [review]

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala (hunk 6 above)
Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:1024]: I'll need to change `specificTest` to a `VarRef => Tree`, but I think that is indeed better.

**Comment by gzm0** [comment]
Updated. I believe I have addressed all the comments. As an addition, I used the new overload of `js.If#apply` (without `elsep`) where applicable.

**Review by sjrd — APPROVED** [review]
LGTM, although this will need a rebase once #4393 is merged.

**Comment by sjrd** [comment]
```
RUNNING timeout -k 5 45m sbt set scalaJSStage in Global := FullOptStage set scalaJSLinkerConfig in helloworld.v2_11 ~= (_.withPrettyPrint(true)) ++2.11.12 helloworld2_11/run helloworld2_11/clean
OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=512M; support was removed in 8.0
OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=512M; support was removed in 8.0
[info] Loading settings for project scalajs-main_pr-4382-2-build from build.sbt ...
[info] Loading project definition from /localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4382@2/project
[info] Loading settings for project scalajs from build.sbt ...
[info] Resolving key references (60875 settings) ...
[info] Detected JDK version 8
[info] Set current project to Scala.js (in build file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4382@2/)
[info] Defining Global / scalaJSStage
[info] The new value will be used by helloworld2_11 / Compile / scalaJSLinkedFile, helloworld2_11 / Compile / scalaJSLinkerResult and 341 others.
[info] 	Run `last` for details.
[info] Reapplying settings...
[info] Detected JDK version 8
[info] Set current project to Scala.js (in build file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4382@2/)
[info] Defining helloworld2_11 / scalaJSLinkerConfig
[info] The new value will be used by helloworld2_11 / Compile / fastLinkJS / scalaJSLinkerConfig, helloworld2_11 / Compile / fastOptJS and 3 others.
[info] 	Run `last` for details.
[info] Reapplying settings...
[info] Detected JDK version 8
[info] Set current project to Scala.js (in build file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4382@2/)
[info] Setting Scala version to 2.11.12 on 33 projects.
[info] Excluded 69 projects, run ++ 2.11.12 -v for more details.
[info] Reapplying settings...
[info] Detected JDK version 8
[info] Set current project to Scala.js (in build file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4382@2/)
[info] Full optimizing /localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-4382@2/examples/helloworld/.2.11/target/scala-2.11/helloworld-opt
[error] java.util.NoSuchElementException: key not found: ClassName<java.lang.ClassCastException>
[error] 	at scala.collection.MapLike.default(MapLike.scala:231)
[error] 	at scala.collection.MapLike.default$(MapLike.scala:230)
[error] 	at scala.collection.AbstractMap.default(Map.scala:59)
[error] 	at scala.collection.mutable.HashMap.apply(HashMap.scala:65)
[error] 	at org.scalajs.linker.backend.emitter.KnowledgeGuardian$KnowledgeAccessor.getModule(KnowledgeGuardian.scala:211)
[error] 	at org.scalajs.linker.backend.emitter.VarGen.foldSameModule(VarGen.scala:199)
[error] 	at org.scalajs.linker.backend.emitter.VarGen.globalVar(VarGen.scala:49)
[error] 	at org.scalajs.linker.backend.emitter.SJSGen.genScalaClassNew(SJSGen.scala:337)
[error] 	at org.scalajs.linker.backend.emitter.CoreJSLib$CoreJSLibBuilder.$anonfun$defineRuntimeFunctions$1(CoreJSLib.scala:447)
[error] 	at org.scalajs.linker.backend.emitter.CoreJSLib$CoreJSLibBuilder.defineFunction2(CoreJSLib.scala:1765)
[error] 	at org.scalajs.linker.backend.emitter.CoreJSLib$CoreJSLibBuilder.defineRuntimeFunctions(CoreJSLib.scala:445)
[error] 	at org.scalajs.linker.backend.emitter.CoreJSLib$CoreJSLibBuilder.buildPreObjectDefinitions(CoreJSLib.scala:127)
[error] 	at org.scalajs.linker.backend.emitter.CoreJSLib$CoreJSLibBuilder.build(CoreJSLib.scala:116)
[error] 	at org.scalajs.linker.backend.emitter.CoreJSLib$.build(CoreJSLib.scala:36)
[error] 	at org.scalajs.linker.backend.emitter.Emitter$CoreJSLibCache.build(Emitter.scala:691)
[error] 	at org.scalajs.linker.backend.emitter.Emitter.$anonfun$emitOnce$6(Emitter.scala:203)
[error] 	at scala.collection.immutable.List.map(List.scala:282)
[error] 	at org.scalajs.linker.backend.emitter.Emitter.$anonfun$emitOnce$5(Emitter.scala:191)
[error] 	at org.scalajs.logging.Logger.time(Logger.scala:42)
[error] 	at org.scalajs.logging.Logger.time$(Logger.scala:40)
[error] 	at org.scalajs.sbtplugin.Loggers$SbtLoggerWrapper.time(Loggers.scala:19)
[error] 	at org.scalajs.linker.backend.emitter.Emitter.emitOnce(Emitter.scala:191)
[error] 	at org.scalajs.linker.backend.emitter.Emitter.emitAvoidGlobalClash(Emitter.scala:151)
[error] 	at org.scalajs.linker.backend.emitter.Emitter.emitInternal(Emitter.scala:127)
[error] 	at org.scalajs.linker.backend.emitter.Emitter.emit(Emitter.scala:79)
[error] 	at org.scalajs.linker.backend.closure.ClosureLinkerBackend.$anonfun$emit$2(ClosureLinkerBackend.scala:85)
[error] 	at org.scalajs.logging.Logger.time(Logger.scala:42)
[error] 	at org.scalajs.logging.Logger.time$(Logger.scala:40)
[error] 	at org.scalajs.sbtplugin.Loggers$SbtLoggerWrapper.time(Loggers.scala:19)
[error] 	at org.scalajs.linker.backend.closure.ClosureLinkerBackend.emit(ClosureLinkerBackend.scala:85)
[error] 	at org.scalajs.linker.standard.StandardLinkerImpl.$anonfun$link$2(StandardLinkerImpl.scala:47)
[error] 	at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:303)
[error] 	at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:37)
[error] 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
[error] 	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
[error] 	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
[error] 	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
[error] 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
[error] 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
[error] (helloworld2_11 / Compile / fullLinkJS) java.util.NoSuchElementException: key not found: ClassName<java.lang.ClassCastException>
[error] Total time: 17 s, completed Jan 21, 2021 2:14:01 PM
```

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala [hunk 12]
```diff
@@ -442,117 +431,92 @@ private[emitter] object CoreJSLib {
...
+              BracketSelect(underlying, i) :=
+                genCallHelper("newArrayObjectInternal", subArrayClassData, lengths, subLengthIndex)
+            })
+          )),
+          Return(result)
+        )
       }
+    )
 
+    private def defineObjectGetClassFunctions(): Tree = {
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:513]: This method was introduced, but it's never called. That probably explains the issue.

**Comment by gzm0** [comment]
OK. So there are two issues here:

- `defineObjectGetClassFunctions` was factored out and not called. Easy to fix :)
- `condBlock` evaluates its arguments eagerly; this means unneeded trees are built, which tries to look up classes that are not in the symbol requirements. I need to think how to fix that (Scala doesn't seem to support by name varargs).

**Comment by sjrd** [comment]
For `condBlock`, perhaps just call it `condTree` and make it take a `=> Tree`, and if at call site we want a block we write
```scala
condTree(cond)(Block(
  ...
)
```

**Review by sjrd — APPROVED** [review]

**Comment by sjrd** [comment]
Somehow this has affected checksizes (in the good direction). Other than that it seems like the CI will pass. Should this still be draft?

File: project/Build.scala [hunk 13]
```diff
@@ -1676,7 +1676,7 @@ object Build {
             Some(ExpectedSizes(
                 fastLink = 781000 to 782000,
                 fullLink = 148000 to 149000,
-                fastLinkGz = 91000 to 92000,
+                fastLinkGz = 90000 to 91000,
                 fullLinkGz = 36000 to 37000,
```
Comment (sjrd) [inline project/Build.scala:1680]: Erf, the CI says
```
[info] Checksizes: reversi2_12 Scala 2.12.12
[info] fastLink size = 780966 (expected Range 781000 to 782000)
[info] fullLink size = 148377 (expected Range 148000 to 149000)
[info] fastLink gzip size = 91118 (expected Range 90000 to 91000)
[info] fullLink gzip size = 36538 (expected Range 36000 to 37000)
```

**Pull request merged**